c_api/include/taichi/taichi.hpp

```diff
@@ -0,0 +1,652 @@
+// C++ wrapper of Taichi C-API
+#include <list>
+#include <vector>
+#include <string>
+#include "taichi/taichi.h"
+
+namespace ti {
+
+// Token type for half-precision floats.
+struct half {
+  uint16_t _;
+};
+
+namespace detail {
+
+// Template type to data type enum.
+template <typename T>
+struct templ2dtype {};
+template <>
+struct templ2dtype<int8_t> {
+  static const TiDataType value = TI_DATA_TYPE_I8;
+};
+template <>
+struct templ2dtype<int16_t> {
+  static const TiDataType value = TI_DATA_TYPE_I16;
+};
+template <>
+struct templ2dtype<int32_t> {
+  static const TiDataType value = TI_DATA_TYPE_I32;
+};
+template <>
+struct templ2dtype<uint8_t> {
+  static const TiDataType value = TI_DATA_TYPE_U8;
+};
+template <>
+struct templ2dtype<uint16_t> {
+  static const TiDataType value = TI_DATA_TYPE_U16;
+};
+template <>
+struct templ2dtype<uint32_t> {
+  static const TiDataType value = TI_DATA_TYPE_U32;
+};
+template <>
+struct templ2dtype<half> {
+  static const TiDataType value = TI_DATA_TYPE_F16;
+};
+template <>
+struct templ2dtype<float> {
+  static const TiDataType value = TI_DATA_TYPE_F32;
+};
+template <>
+struct templ2dtype<double> {
+  static const TiDataType value = TI_DATA_TYPE_F64;
+};
+
+template <typename THandle>
+THandle move_handle(THandle &handle) {
+  THandle out = std::move(handle);
+  handle = TI_NULL_HANDLE;
+  return out;
+}
+
+}  // namespace detail
+
+class Memory {
+  TiRuntime runtime_{TI_NULL_HANDLE};
+  TiMemory memory_{TI_NULL_HANDLE};
+  bool should_destroy_{false};
+
+ public:
+  constexpr bool is_valid() const {
+    return runtime_ != nullptr;
+  }
+  inline void destroy() {
+    if (should_destroy_) {
+      ti_free_memory(runtime_, memory_);
+      memory_ = TI_NULL_HANDLE;
+      should_destroy_ = false;
+    }
+  }
+
+  Memory() {
+  }
+  Memory(const Memory &) = delete;
+  Memory(Memory &&b)
+      : runtime_(detail::move_handle(b.runtime_)),
+        memory_(detail::move_handle(b.memory_)),
+        should_destroy_(std::exchange(b.should_destroy_, false)) {
+  }
+  Memory(TiRuntime runtime, TiMemory memory, bool should_destroy)
+      : runtime_(runtime), memory_(memory), should_destroy_(should_destroy) {
+  }
+  ~Memory() {
+    destroy();
+  }
+
+  Memory &operator=(const Memory &) = delete;
+  Memory &operator=(Memory &&b) {
+    destroy();
+    runtime_ = detail::move_handle(b.runtime_);
+    memory_ = detail::move_handle(b.memory_);
+    should_destroy_ = std::exchange(b.should_destroy_, false);
```

Comment:
Can you try if this way of using `std::exchange` also works after C++20? Looks like the result has to be constexpr after C++20: 

![2022-08-31 09-31-08 的屏幕截图](https://user-images.githubusercontent.com/22334008/187573060-e5b3d277-9825-4ee8-a5e7-bae3a2f946a9.png)

Otherwise I'd suggest not using unstable std functions since std::exchange doesn't seem like doing something super hard or fancy.

Reply:
`std::exchange` becomes an `constexpr` since  C++ 20 and nothing else changes (See the second entry in your screenshot). It is a preferred way to `std::move` a POD type. Please consult this CppCon talk https://youtu.be/ZG59Bqo7qX4?t=1310 for more detail.